### ISSUE TITLE
[package require] fails after [package forget]

### ISSUE THREAD

**Issue by pooryorick** [flightaware/tohil#52]
The following script,

```tcl
package require tohil
package forget tohil
package require tohil
```

returns the error

> attempt to provide package tohil 4.1.0 failed: no version of package tohil provided

. This happens because tohil currently relies on `Tcl_PkgProvide()`, but
`[load]` doesn't call the initialization function a second time.

**Comment by lehenbauer** [comment]
Thanks for that extra explanation, @pooryorick ... it's helpful.  Do you have a suggestion for what we can or should do to fix it?

**Comment by pooryorick** [comment]
You're welcome, and thanks for putting Tohil together.  `Tcl_PkgProvide()` could be replaced by `package provide` in either `pkgIndex.tcl` or in `tohil.tcl`.

**Comment by lehenbauer** [comment]
Cool, doing it in pkgIndex.tcl.in is logical since the version is already getting autoconf-substituted.

And that works.  But the "forget" doesn't seem to do anything insofar as the tohil namespace and all the commands are still there, and they work.

I had found TIP 100 and gotten a Tohil_Unload C function partially working.  unloading the library works but the commands are still there and it's a segfault to try invoke them.  So I guess I would need to delete the commands.  And confirmed, that works.

Do you know if the deal is that the package manager just isn't smart enough to call unload?

**Comment by pooryorick** [comment]
I think `package forget` was designed only to remove the record that the package was provided, not to undo the effects of providing a package.  Probably the idea was to facilitate refreshing code in place, not going to the trouble to delete commands, remove namespaces, etc.   Similarly, a `package unload` procedure was probably considered overkill, as packages could themselves provide a procedure to do that if desired.  It would be nice to have a `package unload` though, along with a way for a package to register an unload procedure for the package.

The fact that `package forget` allows `package require` to call `package ifneeded` multiple times means that a package implementation must be careful to make sure that any initialization the `package ifneeded` script performs doesn't step on the toes of initialization that took place previously.  For example, the sha256 library in Tcl had a problem where the initialization script renamed a procedure, and a subsequent initialization resulted in an error, as the procedure was already renamed.

On a related note, it can be useful to separate [package require] from initialization by providing a separate "init" procedure that can be called after the user loads and configures the package.  Tohil might benefit from such a design.  When I was investigating the issue of Python not finding the Tohil module when Tohil was loaded in an embedded Tcl, there was no way to provide a preliminary Python script to be run prior to importing Tohil.  In the end, the fix didn't require such configuration of the Python interpreter, but it might be a nice thing to have in the future.

There's talk among the Tcl core developers that the ability to unload a shared object from the process ought to be considered a misfeature because there is no way to reliable ensure that all handles to things in the shared object have been released.  I like the idea of being able to unload well-behaved shared objects, and there is an example in the Tcl test suite of how to at least track and unload handles the library itself has control over.

**Comment by lehenbauer** [comment]
Fixed by PR #55
